include/pybind11/typing.h

```diff
@@ -78,6 +78,29 @@ class Optional : public object {
     using object::object;
 };
 
+#if defined(__cpp_nontype_template_parameter_class)
+template <size_t N>
+struct StringLiteral {
+    // This struct is different than detail::descr since it accepts a string rather than a type.
+    // StringLiteral is used in TypeVar to create "instances" of c++ type objects.
+    constexpr StringLiteral(const char (&str)[N]) {
+        // Ensures This copy is done at compile time
```

Comment:
I'd probably delete the comment and the `if` here, too. I feel the `constexpr` explains everything.

To validate what's on my mind:

Adding

```
static_assert(std::is_constant_evaluated(), "");
```

seems redundant. (?)